### DISCUSSION TITLE
How to apply symmetry boundary conditions?

### DISCUSSION THREAD

**Discussion by RWTHLHK** [idaholab/moose#23673] [Q&A Modules: Solid mechanics] [answered]
I need to run a simulation with 1/8 3D mesh. So I need to apply symmetry boundary conditions. I know the symmetry boundary condition in x-direction is (disp_x = rot_y = rot_z = 0), but when I add rotations in GlobalParams by rotations = "rot_x rot_y rot_z", it says the variables "rot_x, rot_y and rot_z" can't be found. And if I only set disp_x to 0, the simulation results are totally wrong. The attached file is  my input file.
[problems.zip](https://github.com/idaholab/moose/files/10930933/problems.zip)

**Comment by GiudGiud** [comment]
Hello

You should either create the rotation variables or find the Action that does that for you
@recuero @jiangwen84  do you know

Guillaume

**Comment by recuero** [comment, marked as answer]
I'd just add the rotation variables to the pertinent blocks --not to `GlobalParams`. That's the way most input files are set up as far as I can see.

**Reply by RWTHLHK** [reply]
Would you please explain this in more detail? I just looked at the examples of tensor mechanic module and couldn't find a relevant tutorial.

**Reply by recuero** [reply]
I cannot see any rotations in your attached problem and am not sure what you are trying to simulate or which elements you want to use. As @GiudGiud pointed out, the rotation variables can be created by actions. For rotation variables necessary for beams/trusses, the `LineElementMaster` action will do this for you. If you are wanting to use some other element you will need to create those variables and pass them appropriately to compute internal forces and apply BCs. For general tutorials, refer to https://mooseframework.inl.gov/getting_started/examples_and_tutorials/index.html. For particular examples on how to use variables in the tensor mechanics modules, you can consult https://mooseframework.inl.gov/modules/tensor_mechanics/index.html and browse relevant input files.

**Reply by RWTHLHK** [reply]
Hello, thanks a lot for your reply! I need to apply boundary conditions to a model as shown in the following figure, which is 1/8 of a complete model. The symmetry boundary conditions should be applied to x-face(the face normal to x-direction) and z-face(the face normal to z-direction), which are named as inside1 and inside2 in my input file, respectively. Because it is an uni-axial tensile loading, I only set displacements of these faces to 0, but the results look weird.
`[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
    [base]
        type = FileMeshGenerator
        file = NRB_R2.inp
    []
[]

[Modules/TensorMechanics/Master]
  [./all]
    strain = FINITE
    add_variables = true
    generate_output = 'max_principal_stress triaxiality_stress vonmises_stress plastic_strain_xx plastic_strain_yy plastic_strain_zz'
  [../]
[]

[Functions]
  [./hf]
    type = PiecewiseLinear
    data_file = stress_data.csv
    format = columns
  [../]
[]

[BCs]
  [./y_pull_function]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = 0.3*t
  [../]
  [./xbot]
    type = DirichletBC
    variable = disp_x
    boundary = inside1
    value = 0.0
  [] 
  [./y_bot]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  [../]
  [./zbot]
    type = DirichletBC
    variable = disp_z
    boundary = inside2
    value = 0.0
  []
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2e11#Pa
    poissons_ratio = 0.29
  [../]
  [./isotropic_plasticity]
    type = IsotropicPlasticityStressUpdate
    yield_stress = 431.46e6#Pa
    hardening_function = hf
  [../]
  [./radial_return_stress]
    type = ComputeMultipleInelasticStress
    tangent_operator = elastic
    inelastic_models = 'isotropic_plasticity'
  [../]
  [density]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 7850
  []
  [lode_angle]
    type = ComputeLodeAngle
    outputs = exodus
  []
  [critical_crack_strain]
    type = ComputeCriticalCrackStrain
    outputs = exodus
  []
  [peeq]
    type = ComputePEEQ
    outputs = exodus
  []
  [d]
    type = ComputeDamage
    alpha = 1.0
    beta = 1.0
    critical_fracture_stress = 1204
    outputs = exodus
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'

  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-ksp_gmres_restart'
  petsc_options_value = '101'

  line_search = 'none'

  l_max_its = 50
  nl_max_its = 50
  nl_rel_tol = 1e-10
  nl_abs_tol = 1e-12
  l_tol = 1e-9

  start_time = 0.0
  end_time = 0.5
  dt = 0.005
[]

[Outputs]
  exodus = true
  console = true
  print_perf_log = true
[]

`
![NRB_part](https://user-images.githubusercontent.com/43910101/224295248-8042ac41-c82d-40d0-bc51-8043b0e1548c.png)

**Reply by recuero** [reply]
Solid elements, with rare exceptions, do not have rotational variables. I don't think that the cause of the weirdness in your results is caused by rotational degrees of freedom. The BCs, as you describe them, look correct to me.

You have other objects in your input file, such as `ComputeLodeAngle` that I am not familiar with, nor can I find them in MOOSE. Could you maybe simplify your constitutive model and rerun the problem? You may see if your results make more sense with mere elastic behavior.
